sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1347,20 +1347,84 @@ void gnrc_netif_default_init(gnrc_netif_t *netif)
 #endif
 }
 
+#ifdef MODULE_GNRC_NETIF_EVENTS
+/**
+ * @brief   Call the ISR handler from an event
+ *
+ * @param[in]   evp     pointer to the event
+ */
+static void _event_handler_isr(event_t *evp)
+{
+    gnrc_netif_t *netif = container_of(evp, gnrc_netif_t, event_isr);
+    netif->dev->driver->isr(netif->dev);
+}
+#endif /* MODULE_GNRC_NETIF_EVENTS */
+
+/**
+ * @brief   Process any pending events and wait for IPC messages
+ *
+ * This function will block until an IPC message is received. Events posted to
+ * the event queue will be processed while waiting for messages.
+ *
+ * @param[in]   netif   gnrc_netif instance to operate on
+ * @param[out]  msg     pointer to message buffer to write the first received message
+ *
+ * @return >0 if msg contains a new message
+ */
+static void _process_events_await_msg(gnrc_netif_t *netif, msg_t *msg)
+{
+#ifdef MODULE_GNRC_NETIF_EVENTS
+    while (1) {
+        /* Using messages for external IPC, and events for internal events */
+
+        /* First drain the queues before blocking the thread */
+        /* Events will be handled before messages */
+        DEBUG("gnrc_netif: handling events\n");
+        event_t *evp;
+        /* We can not use event_loop() or event_wait() because then we would not
+         * wake up when a message arrives */
+        while ((evp = event_get(&netif->evq))) {
+            DEBUG("gnrc_netif: event %p\n", (void *)evp);
+            if (evp->handler) {
+                evp->handler(evp);
+            }
+        }
+        /* non-blocking msg check */
+        int msg_waiting = msg_try_receive(msg);
+        if (msg_waiting > 0) {
+            return;
+        }
+        DEBUG("gnrc_netif: waiting for events\n");
+        /* Block the thread until something interesting happens */
+        thread_flags_wait_any(THREAD_FLAG_MSG_WAITING | THREAD_FLAG_EVENT);
+    }
+#else /* MODULE_GNRC_NETIF_EVENTS */
+    (void) netif;
+    /* Only messages used for event handling */
+    DEBUG("gnrc_netif: waiting for incoming messages\n");
+    msg_receive(msg);
+#endif /* MODULE_GNRC_NETIF_EVENTS */
+}
+
 static void *_gnrc_netif_thread(void *args)
 {
     gnrc_netapi_opt_t *opt;
     gnrc_netif_t *netif;
     netdev_t *dev;
     int res;
     msg_t reply = { .type = GNRC_NETAPI_MSG_TYPE_ACK };
-    msg_t msg, msg_queue[CONFIG_GNRC_NETIF_MSG_QUEUE_SIZE];
+    msg_t msg_queue[CONFIG_GNRC_NETIF_MSG_QUEUE_SIZE];
 
     DEBUG("gnrc_netif: starting thread %i\n", sched_active_pid);
     netif = args;
     gnrc_netif_acquire(netif);
     dev = netif->dev;
     netif->pid = sched_active_pid;
+#ifdef MODULE_GNRC_NETIF_EVENTS
```

Comment:
why don't we just make it mandatory? Since this change only affects gnrc_netif, I see absolutely no reason to continue using messages to process the IRQ

Reply:
Could be, but that's something we have to decide on as enabling this by default would increase the RAM and ROM usage of all gnrc_netif applications.

Reply:
We can always enable this by default in the future.

Reply:
Yes, but when using messages interrupts can be lost. In some cases this makes the device unresponsive (`at86rf2xx`). To me this is actually a fix, besides being a cool feature

Reply:
@miri64 Do you have an opinion on enabling this by default?

Reply:
I have nothing against enabling by default, but I would prefer to keep it optional for now.

Reply:
ok, fine with me